### PR TITLE
[OSX] Allow executing the bundle to return pid

### PR DESCRIPTION
**Pull request by oneeyeman1**
This PR will take care of the ticket 9603 and 14883.

After applying you will receive proper pid after asynchronously executing OSX bundle.

The test suite does not test this functionality under OSX.

This code will need to be re-written for the next version of OSX (10.15), where the API used in this PR becomes deprecated.

However, I don't have the next version installed and can't test the new OSX. So this PR does not include support for the 10.15 Beta release.

The sample seems to work and it does return the pid of the running Bundle.

The code has been tested under 10.13 with the minimum deployment version of 10.9.

Please review and apply.

### CODE REVIEW

**Comment by oneeyeman1** [comment]
@csomor ,
Stefan, any word about that?

Thank you.

**Comment by csomor** [comment]
I'm not really the specialist here, but the code seems to follow the same lines as Ryan's Version did, so if the sample works correctly, I'm fine with applying it, in order to get further testing coverage. Thanks.

File: src/osx/cocoa/utils_base.mm [hunk 1]
```diff
@@ -191,3 +191,95 @@ wxString wxGetOsDescription()
     return true;
 }
 #endif // wxUSE_DATETIME
+
+bool wxCocoaLaunch(const char* const* argv, pid_t &pid)
+{
+    // Obtains the number of arguments for determining the size of
+    // the CFArray used to hold them
+    NSUInteger cfiCount = 0;
```
Comment (vadz) [inline src/osx/cocoa/utils_base.mm]: This is not used except for checking if it's 0, which could be done with just `if ( !argv )`

File: src/osx/cocoa/utils_base.mm [hunk 2]
```diff
@@ -191,3 +191,95 @@ wxString wxGetOsDescription()
...
+    // turn them into CFURLs and then put them in cfaFiles
+    // For use to launch services call
+    NSMutableArray *params = [[NSMutableArray alloc] init];
+    for( ; *argv != NULL; ++argv )
+    {
+        // Check for '<' as this will ring true for
+        // CFURLCreateWithString but is generally not considered
+        // typical on mac but is usually passed here from wxExecute
+        if (wxStrcmp(*argv, wxT("<")) == 0)
+            continue;
```
Comment (vadz) [inline src/osx/cocoa/utils_base.mm]: This doesn't make any sense to me, why should we silently ignore `<`? Of course, it shouldn't be here in the first place (this is `wxExecute()`, not `wxShell()`), but this entire check should be just removed.

File: src/osx/cocoa/utils_base.mm [hunk 3]
```diff
@@ -191,3 +191,95 @@ wxString wxGetOsDescription()
...
+        {
+            wxLogDebug(
+                       wxT("wxCocoaLaunch Could not create NSURL for argument:%s"),
+                       *argv);
+            continue;
+        }
+        
+        // Add the valid CFURL to the argument array and then
+        // release it as the CFArray adds a ref count to it
+        [params addObject:cfurlCurrentFile];
```
Comment (vadz) [inline src/osx/cocoa/utils_base.mm]: I'm not sure about Cocoa ownership rules... Are we sure we don't leak `cfurlCurrentFile` here?

File: src/osx/cocoa/utils_base.mm [hunk 4]
```diff
@@ -191,3 +191,95 @@ wxString wxGetOsDescription()
...
+    NSWorkspace *ws = [NSWorkspace sharedWorkspace];
+    NSRunningApplication *app = [ws launchApplicationAtURL:url options:NSWorkspaceLaunchAsync
+                                             configuration:[NSDictionary dictionaryWithObject:params forKey:NSWorkspaceLaunchConfigurationArguments]
+                                             error:&error];
+    if( app != nil )
+        pid = [app processIdentifier];
+    else
+    {
+        wxString errorDesc = wxCFStringRef::AsString([error localizedDescription]);
+        wxLogDebug( wxString::Format( "wxCocoaLaunch failure: error is %s", errorDesc ) );
```
Comment (vadz) [inline src/osx/cocoa/utils_base.mm]: `Format()` inside `wxLogXXX()` is useless.

File: src/unix/utilsunx.cpp [hunk 5]
```diff
@@ -585,15 +585,16 @@ long wxExecute(const char* const* argv, int flags, wxProcess* process,
...
+    pid_t pid;
 #if defined(__DARWIN__) && !defined(__WXOSX_IPHONE__)
+    pid = -1;
     // wxMacLaunch() only executes app bundles and only does it asynchronously.
     // It returns false if the target is not an app bundle, thus falling
     // through to the regular code for non app bundles.
-    if ( !(flags & wxEXEC_SYNC) && wxMacLaunch(argv) )
+    if ( !(flags & wxEXEC_SYNC) && wxCocoaLaunch(argv, pid) )
     {
         // we don't have any PID to return so just make up something non null
```
Comment (vadz) [inline src/unix/utilsunx.cpp]: This comment is out of date.

**Review by vadz — COMMENTED** [review]
Sorry, I hadn't had time to review this, but there are a few things that it would be really nice to fix here.

File: src/osx/cocoa/utils_base.mm (hunk 2 above)
Comment (csomor) [reply src/osx/cocoa/utils_base.mm]: these lines and their comment comes from Ryans version from way back when, so ok, let's remove that

Comment (oneeyeman1) [reply src/osx/cocoa/utils_base.mm]: @vadz ,
I guess the idea is that OSX bundle can't take input from the command line.

I will see if removing those will produce the same result.

File: src/unix/utilsunx.cpp (hunk 5 above)
Comment (oneeyeman1) [reply src/unix/utilsunx.cpp]: Oops, sorry. You are right.
Will remove.

**Comment by oneeyeman1** [comment]
@vadz ,
How will it work?
I see that you merged this PR, so what will happen when I fix the things from the review ?

Thank you.

**Comment by csomor** [comment]
I've merged it, I'll go through the comments and commit the changes in one batch

**Comment by oneeyeman1** [comment]
@csomor 
> I've merged it, I'll go through the comments and commit the changes in one batch

Thx.

File: src/osx/cocoa/utils_base.mm (hunk 3 above)
Comment (csomor) [reply src/osx/cocoa/utils_base.mm]: we leak params, I'll fix that

File: src/osx/cocoa/utils_base.mm (hunk 1 above)
Comment (csomor) [reply src/osx/cocoa/utils_base.mm]: I'll do that

File: src/osx/cocoa/utils_base.mm (hunk 2 above)
Comment (csomor) [reply src/osx/cocoa/utils_base.mm]: ok, I'm waiting on this, if it's fine without, I'll commit the changes

File: src/osx/cocoa/utils_base.mm (hunk 4 above)
Comment (csomor) [reply src/osx/cocoa/utils_base.mm]: done

**Pull request merged**